### PR TITLE
SQL Datasources: Allow multiple string/labels columns with time series

### LINKED ISSUE #36370: Labels for SQL data sources based on additional returned columns
**What would you like to be added**:
Grafana should treat additional columns returned by a query against a sql data source (e.g. postgres) as key/value pairs and allow them to be used as regular labels.

``` sql
SELECT
  captured AS time,
  name AS metric,
  city, -- additional key/value label {city=...}
  speed AS value
FROM car_status
```

time (captured)|metric (name)|city|value (speed)
--|--|--|--
2021-07-02T08:50:00|car1|london|100
2021-07-02T08:50:00|car2|berlin|150

**Why is this needed**:
Some data sources (e.g. prometheus) provide labels out-of-the-box and they are handled by grafana as-it. Sql data sources like postgres do not provide labels, you only got your query which returns columns. This becomes a **really big issue** when you try to use the new alerting system of grafana 8 - as it relies hevily on them and without labels it is kind of not useable.

References pr #32257, bug #32899.

### PR DESCRIPTION
**Pull request by kylebrandt**
**What this PR does / why we need it**:
Allows "Long" format with multiple string columns, currently if you select more than one string columns it seems the code tries to convert these columns into a numeric field instead of labels.

It works with this PR:

image

Before I would get the error:

`convert value to float failed: metricIndex 1 type can't be converted to float`

**Which issue(s) this PR fixes**:

Fixes  #36370

**Special notes for your reviewer**:
How the backwards compatiability works confuses me some - may need to check I did not break that.

### CODE REVIEW

**Comment by marefr** [comment]
Perhaps somewhat related to #35876 ?

File: pkg/tsdb/sqleng/sql_engine.go [hunk 1]
```diff
@@ -288,6 +288,10 @@ func (e *dataPlugin) executeQuery(query plugins.DataSubQuery, wg *sync.WaitGroup
 				continue
 			}
 
+			if t := frame.Fields[i].Type(); t == data.FieldTypeString || t == data.FieldTypeNullableString {
```
Comment (marefr) [inline pkg/tsdb/sqleng/sql_engine.go:291]: Maybe this should be handled in `newProcessCfg`? Change `metricIndex` to `metricIndices` perhaps? On the other hand the logic for finding one metricIndex might no longer be applicable with your suggested changes?

**Review by marefr — COMMENTED** [review]
Tested this quickly and seems to continue work as earlier for queries only having one metric column. One thing that gets a bit tricky with this is legend names/aliases - how can you format that? I guess you have to do it in the SQL query?

**Comment by kylebrandt** [comment]
> Tested this quickly and seems to continue work as earlier for queries only having one metric column. One thing that gets a bit tricky with this is legend names/aliases - how can you format that? I guess you have to do it in the SQL query?

In the Long format when (which gets triggered on multiple strings) any numeric column is a value column (and there can be, but doesn't have to be, multiple numeric columns). So by default, the metric portion of the name is that column name. To format that one could either do it in the query with "AS" or use Display Name and/or field overrides.

**Comment by kylebrandt** [comment]
@marefr / @ying-jeanne any thoughts on merging this?

**Comment by marefr** [comment]
Should be fine I think, but issue labeled as feature so think this is more suitable for 8.1.0. WDYT?

**Comment by kylebrandt** [comment]
> Should be fine I think, but issue labeled as feature so think this is more suitable for 8.1.0. WDYT?

Updated the tags to 8.1

**Comment by marefr** [comment]
> Updated the tags to 8.1

@kylebrandt Go for it :+1:

**Review by marefr — APPROVED** [review]

**Pull request merged**